tests/integration/relation_tests/application-charm/lib/charms/data_platform_libs/v0/database_requires.py

```diff
@@ -0,0 +1,312 @@
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.database = database_name
+        self.extra_user_roles = extra_user_roles
+        self.local_app = self.charm.model.app
+        self.local_unit = self.charm.unit
+        self.relation_name = relation_name
+        self.framework.observe(
+            self.charm.on[relation_name].relation_joined, self._on_relation_joined_event
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_changed, self._on_relation_changed_event
+        )
+
+    def _diff(self, event: RelationChangedEvent) -> Diff:
+        """Retrieves the diff of the data in the relation changed databag.
+
+        Args:
+            event: relation changed event.
+
+        Returns:
+            a Diff instance containing the added, deleted and changed
+                keys from the event relation databag.
+        """
+        # Retrieve the old data from the data key in the application relation databag.
+        old_data = json.loads(event.relation.data[self.charm.model.app].get("data", "{}"))
+        # Retrieve the new data from the event relation databag.
+        new_data = {
+            key: value for key, value in event.relation.data[event.app].items() if key != "data"
+        }
+
+        # These are the keys that were added to the databag and triggered this event.
+        added = new_data.keys() - old_data.keys()
+        # These are the keys that were removed from the databag and triggered this event.
+        deleted = old_data.keys() - new_data.keys()
+        # These are the keys that already existed in the databag,
+        # but had their values changed.
+        changed = {
+            key for key in old_data.keys() & new_data.keys() if old_data[key] != new_data[key]
+        }
+
+        # TODO: evaluate the possibility of losing the diff if some error
+        # happens in the charm before the diff is completely checked (DPE-412).
+        # Convert the new_data to a serializable format and save it for a next diff check.
+        event.relation.data[self.local_app].update({"data": json.dumps(new_data)})
+
+        # Return the diff with all possible changes.
+        return Diff(added, changed, deleted)
+
+    def fetch_relation_data(self) -> dict:
+        """Retrieves data from relation.
+
+        This function can be used to retrieve data from a relation
+        in the charm code when outside an event callback.
+
+        Returns:
+            a dict of the values stored in the relation data bag
+                for all relation instances (indexed by the relation id).
+        """
+        data = {}
+        for relation in self.relations:
+            data[relation.id] = {
+                key: value for key, value in relation.data[relation.app].items() if key != "data"
+            }
+        return data
+
+    def _update_relation_data(self, relation_id: int, data: dict) -> None:
```

Comment:
Hey, I don't know jack about Python really, but I do know Java. In Java, this method would typically be prefixed with `private` to denote it's scope, which is pretty evidently limited to the class. In Python, I understand that declaring private scope is done by prefixing a double underscore to the method name. Just wondering why we are not using that convention for this stuff?

Reply:
maybe something to do with the `self.framework.observe()` stuff triggering this event? If so feel free to ignore this comment!

Reply:
Hi @grobbie ! I can't take credit for this awesome code, it was merely ported from [database-libs](). So I may not be answering this question perfectly. But I think on the DPE team we have been using a single underscore to denote private functions. 

I will admit that I don't know much about the knitty-gritty syntax of Python since I am a somewhat new professional, but I think it might be a contested point that sometimes private functions use `__` and sometimes they use `_`. [This stackoverflow post](https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name#:~:text=Single%20underscore%20at%20the%20beginning,not%20part%20of%20the%20API.) explains the difference between the two, but I have seen that double underscore can be used for private (but again on the DPE team we've been using a single underscore)